apis/v1alpha2/shared_types.go

```diff
@@ -401,3 +430,86 @@ type AnnotationKey string
 // +kubebuilder:validation:MinLength=0
 // +kubebuilder:validation:MaxLength=4096
 type AnnotationValue string
+
+// AddressRouteMatches defines AddressMatch rules for inbound traffic according to
+// source and/or destination address of a packet or connection.
+type AddressRouteMatches struct {
+	// SourceAddresses indicates the originating (source) network
+	// addresses which are valid for routing traffic.
+	//
+	// Support: Extended
+	SourceAddresses []AddressMatch `json:"sourceAddresses"`
+
+	// DestinationAddresses indicates the destination network addresses
+	// which are valid for routing traffic.
+	//
+	// Support: Extended
+	DestinationAddresses []AddressMatch `json:"destinationAddresses"`
```

Comment:
How is this different than listener?  Is this related to SNI?  No?

I'd like to understand how this is used - hostname isn't part of a TCP or UDP packet.  Is someone planning to do DNS lookups and filter on that?

Reply:
I think there are at least two key use cases here:

1. For routing requests going through Gateway(s) that are listening on a wide set of addresses
2. For routing requests not going through Gateway(s) such as mesh implementations

I think @shaneutt and/or @howardjohn might be able to add some more detail for potential use cases here.

Reply:
The strongest use case, IMO, is for east-west traffic that can convey a destination address. This could be from CONNECT, PROXY protocol, service mesh redirection, etc.

For example, if we want to say "All requests to httpbin should use canary logic":

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: HTTPRoute
metadata:
  name: httpbin-canary
spec:
  parentRefs:
  - name: my-gateway
  matches:
  - destinationAddresses:
      type: NamedAddress
      value: httpbin.default.svc.cluster.local
  rules:
  - backendRefs:
    - name: httpbin-v2
      port: 80
      weight: 1
    - name: httpbin-v1
      port: 80
      weight: 1
```

(note: HTTP doesn't actually have this field currently, but the same applies to all protocols).

Matching only on Host header is not actually sufficient, IMO. For example, consider if we are trying to configure sending traffic to an nginx proxy, which then in turn does Host header routing (that is, the application does `curl nginx.default.svc.cluster.local -H "Host: host-for-nginx"`. Matching `hostnames: [nginx.default.svc.cluster.local]` would not help here.

Reply:
Discussed in community meeting and it seems like we may need to rethink this a bit, more discussion in #1127.

Reply:
This has been removed from the API while we reconsider how it could work.